[filebeat][httpjson] - adding support for replace_with clause & first_response context object (#33373)
elastic/beats: x-pack/filebeat/docs/inputs/input-httpjson.asciidoc

+
Collect and make events from response in any format supported by httpjson for all calls.

+
[[chain-step-replace_with]]
[float]
==== `chain[].step.replace_with`

> The `replace_with: "pattern,value"` clause is used to replace a pattern defined in the `request.url` with a given value. The value can be hard coded or can be 

Author: Agreed, but one concern I have is, in our case we are not replacing any string, but specific strings with a certain structure. Every old_value string in our case needs to start with **$.** which make it more similar to a pattern rather than a normal string. Maybe we could add a **NOTE** saying traditional regex patterns are not supported ?

> The `replace_with: "pattern,value"` clause is used to replace a pattern defined in the `request.url` with a given value. The value can be hard coded or can be 

Reviewer: I'm thinking "fixed pattern" would probably cover the distinction.

+
<<response-json2-replace_with,response>>

["source","yaml",subs="attributes"]
----
filebeat.inputs:
- type: httpjson
  enabled: true
  # first call
  request.url: https://example.com/services/data/v1.0/exports
  interval: 1h
  chain:
    # second call
    - step:
      request.url: https://example.com/services/data/v1.0/$.exportId/files
      request.method: GET
      replace_with: '$.exportId,first_response.body.exportId'
----

Example:

- First call to fetch exportId

+
request_url: \https://example.com/services/data/v1.0/exports

+
response_json:

+
[[response-json1-replace_with]]
["source","json",subs="attributes"]
----
{
    "exportId" : "2212"
}
----

- Second call to fetch `file ids` using exportId from first call.

+
request_url using exportId as '2212': \https://example.com/services/data/v1.0/2212/files

+
response_json using exportId as '2212':

+
[[response-json2-replace_with]]
["source","json",subs="attributes"]
----
{
    "files": [
        {
            "id": 1,
        },
        {
            "id": 2,
        },
        {
            "id": 3,
        },
    ]
}
----
This behaviour of targeted pattern replacement in the url helps solve various use cases.

> NOTE: Replace patterns should not contain commas `","` in their definition.

Author: Doing this would remove the highlights. `` inside " " applies italics to text and does not highlight the text

> NOTE: Replace patterns should not contain commas `","` in their definition.

Reviewer: Single quotes? (aside: that's horrific)

> NOTE: Replace patterns should not contain commas `","` in their definition.

Author: Single quotes also have the same behaviour : Italics & no highlights

> NOTE: Replace patterns should not contain commas `","` in their definition.

Author: the best that can be done is a `"`,`"` .. which will italic the quotes ..suggesting the user that it's not a part of the text

> NOTE: Replace patterns should not contain commas `","` in their definition.

Reviewer: It looks like this is only possible by using a ZWJ between the backticks and the double quotes.

``"\U200D`,`\U200D"``.

> NOTE: Replace patterns should not contain commas `","` in their definition.

Author: It's showing "\U200D`,`\U200D" in my local asciidoc preview window. It's not parsing "\U200D" . I'll just remove the symbol for comma and just highlight the text. It should be enough.

> NOTE: Replace patterns should not contain commas `","` in their definition.

Reviewer: Yeah that's a not a parsable bit of source. On a linux box it would be entered as ctrl-shift-u 200d, and it won't be visible in many editors; so it's not the way to go.

+
Collect and make events from response in any format supported by httpjson for all calls.

+
[[chain-step-replace_with]]
[float]
==== `chain[].step.replace_with`

> The `replace_with: "fixed_pattern,value"` clause is used to replace a fixed pattern string defined in `request.url` with a given value. It is 
> called a fixed pattern string because it always needs to start with a `$.` character, for example: `$.xyz`. The `value` can be hard coded or 
> can be extracted from context variables like [`.last_response.*`, `.first_response.*`] etc. The `replace_with` clause can be used in 
> combination with the `replace` clause thus providing a lot of flexibility in the logic of chain requests.

Reviewer: ```suggestion
The `replace_with: "pattern,value"` clause is used to replace a fixed pattern string defined in `request.url` with the given value. The fixed pattern must have a `$.` prefix, for example: `$.xyz`. The `value` may be hard coded or 
extracted from context variables like [`.last_response.*`, `.first_response.*`] etc. The `replace_with` clause can be used in 
combination with the `replace` clause thus providing a lot of flexibility in the logic of chain requests.
```

+
<<response-json2-replace_with,response>>

["source","yaml",subs="attributes"]
----
filebeat.inputs:
- type: httpjson
  enabled: true
  # first call
  request.url: https://example.com/services/data/v1.0/exports
  interval: 1h
  chain:
    # second call
    - step:
      request.url: https://example.com/services/data/v1.0/$.exportId/files
      request.method: GET
      replace_with: '$.exportId,first_response.body.exportId'
----

Example:

- First call to fetch exportId

+
request_url: \https://example.com/services/data/v1.0/exports

+
response_json:

+
[[response-json1-replace_with]]
["source","json",subs="attributes"]
----
{
    "exportId" : "2212"
}
----

- Second call to fetch `file ids` using exportId from first call.

+
request_url using exportId as '2212': \https://example.com/services/data/v1.0/2212/files

+
response_json using exportId as '2212':

+
[[response-json2-replace_with]]
["source","json",subs="attributes"]
----
{
    "files": [
        {
            "id": 1,
        },
        {
            "id": 2,
        },
        {
            "id": 3,
        },
    ]
}
----
This behaviour of targeted fixed pattern replacement in the url helps solve various use cases.

> NOTE: Fixed patterns must not contain `commas` in their definition. String replacement patterns are matched by the 

Reviewer: I agree this is simpler, but we don't need the pre span on it.

```suggestion
NOTE: Fixed patterns must not contain commas in their definition. String replacement patterns are matched by the 
```

----
This behaviour of targeted pattern replacement in the url helps solve various use cases.

> NOTE: Replace patterns must not contain commas `","` or `regular expressions` in their definition.

Reviewer: I think the added clause here is confusing. Better to not raise this here since it makes the user think that REs are somehow involved and they are no. Calling the pattern in the text above a "fixed pattern", potentially with a final sentence along the lines of "String replacement patterns are matched by the \`replace_with\` processor with exact string matching."

Author: updated the docs accordingly.
